### PR TITLE
Setup deployment and CD

### LINKED ISSUE #4: Setup deployment and CD
- [x] Setup deployment
- [x] Deploy to Heroku

### PR DESCRIPTION
**Pull request by Lahphim**
Resolve https://github.com/Lahphim/go-crawler-challenge/issues/4

## What happened 👀

Setup deployment workflow with Github Action 🙌

## Insight 📝

- [x] Setup deployment
- [x] Deploy to Heroku

For deploy to the Heroku, I'm using this Git Action https://github.com/marketplace/actions/deploy-to-heroku
This package provide an option to deploy with docker also ✌️

## Proof Of Work 📹

CD : https://github.com/Lahphim/go-crawler-challenge/runs/1594120870?check_suite_focus=true
Screen Shot 2563-12-22 at 17 57 24

When deploy to the staging project.
https://go-challenge-staging.herokuapp.com/
Screen Shot 2563-12-22 at 18 10 07

When deploy to the production project.
https://go-challenge.herokuapp.com/
Screen Shot 2563-12-22 at 18 09 43

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+name: Deploy
+
+on:
+  workflow_run:
+    workflows:
+      - Test
```
Comment (sleepylee) [inline .github/workflows/deploy.yml:6]: should this be "Deploy"?

Comment (Lahphim) [reply .github/workflows/deploy.yml:6]: @sleepylee As I set as `Test` because I'll trigger this workflow after `Test` complete ✌️

**Review by olivierobert — APPROVED** [review]

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            echo "HEROKU_APP_NAME=${{ secrets.HEROKU_APP_NAME }}-staging" >> $GITHUB_ENV
+          fi
+
+      - uses: akhileshns/heroku-deploy@v3.7.8
+        with:
+          heroku_api_key: ${{ secrets.HEROKU_API_KEY }}
+          heroku_app_name: $HEROKU_APP_NAME
+          heroku_email: ${{ secrets.HEROKU_APP_NAME }}
+          usedocker: true
+          env_file: ".env"
```
Comment (olivierobert) [inline .github/workflows/deploy.yml:43]: While this `.env` is committed to the repository, here might be future env variables that should remain private. So think of using Github secrets to append these environment variables to an `.env` file as done on the website:

image

Comment (Lahphim) [reply .github/workflows/deploy.yml:43]: @olivierobert Noted

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            echo "HEROKU_APP_NAME=${{ secrets.HEROKU_APP_NAME }}" >> $GITHUB_ENV
+          else
+            echo "HEROKU_APP_NAME=${{ secrets.HEROKU_APP_NAME }}-staging" >> $GITHUB_ENV
+          fi
+
+      - uses: akhileshns/heroku-deploy@v3.7.8
+        with:
+          heroku_api_key: ${{ secrets.HEROKU_API_KEY }}
+          heroku_app_name: $HEROKU_APP_NAME
+          heroku_email: ${{ secrets.HEROKU_APP_NAME }}
```
Comment (hoangmirs) [inline .github/workflows/deploy.yml]: Just curious, why did you use `HEROKU_APP_NAME` for email?

Comment (Lahphim) [reply .github/workflows/deploy.yml]: @hoangmirs My bad 🙌
I copy the wrong one. Actually I have `HEROKU_APP_EMAIL` for that part.

Updated a3614c1bb229120b7c43e3d2184cda6cbc377dc0

Screen Shot 2563-12-23 at 16 55 58

**Review by hoangmirs — APPROVED** [review]

**Pull request merged**